packages/loader/container-loader/src/connectionHandler.ts

```diff
@@ -0,0 +1,886 @@
+} from "./contracts";
+
+const MaxReconnectDelayInMs = 8000;
+const InitialReconnectDelayInMs = 1000;
+const DefaultChunkSize = 16 * 1024;
+
+const fatalConnectErrorProp = { fatalConnectError: true };
+
+function getNackReconnectInfo(nackContent: INackContent) {
+    const message = `Nack (${nackContent.type}): ${nackContent.message}`;
+    const canRetry = nackContent.code !== 403;
+    const retryAfterMs = nackContent.retryAfter !== undefined ? nackContent.retryAfter * 1000 : undefined;
+    return createGenericNetworkError(
+        `nack [${nackContent.code}]`, message, canRetry, retryAfterMs, { statusCode: nackContent.code });
+}
+
+const createReconnectError = (fluidErrorCode: string, err: any) =>
+    wrapError(
+        err,
+        (errorMessage: string) => new GenericNetworkError(fluidErrorCode, errorMessage, true /* canRetry */),
+    );
+
+/**
+ * Implementation of IDocumentDeltaConnection that does not support submitting
+ * or receiving ops. Used in storage-only mode.
+ */
+class NoDeltaStream
+    extends TypedEventEmitter<IDocumentDeltaConnectionEvents>
+    implements IDocumentDeltaConnection, IDisposable
+{
+    clientId: string = "storage-only client";
+    claims: ITokenClaims = {
+        scopes: [ScopeType.DocRead],
+    } as any;
+    mode: ConnectionMode = "read";
+    existing: boolean = true;
+    maxMessageSize: number = 0;
+    version: string = "";
+    initialMessages: ISequencedDocumentMessage[] = [];
+    initialSignals: ISignalMessage[] = [];
+    initialClients: ISignalClient[] = [];
+    serviceConfiguration: IClientConfiguration = {
+        maxMessageSize: 0,
+        blockSize: 0,
+        summary: undefined as any,
+    };
+    checkpointSequenceNumber?: number | undefined = undefined;
+    submit(messages: IDocumentMessage[]): void {
+        this.emit("nack", this.clientId, messages.map((operation) => {
+            return {
+                operation,
+                content: { message: "Cannot submit with storage-only connection", code: 403 },
+            };
+        }));
+    }
+    submitSignal(message: any): void {
+        this.emit("nack", this.clientId, {
+            operation: message,
+            content: { message: "Cannot submit signal with storage-only connection", code: 403 },
+        });
+    }
+
+    private _disposed = false;
+    public get disposed() { return this._disposed; }
+    public dispose() { this._disposed = true; }
+}
+
+/**
+ * Implementation of IConnectionManager, used by Container class
+ * Implements constant connectivity to relay service, by reconnecting in case of loast connection or error.
+ * Exposes various controls to influecen this process, including manual reconnects, forced read-only mode, etc.
+ */
+export class ConnectionManager implements IConnectionManager {
+    // Connection mode used when reconnecting on error or disconnect.
+    private readonly defaultReconnectionMode: ConnectionMode;
+
+    private pendingConnection = false;
```

Comment:
Nice to add comments on all these fields.  `/** blah blah */` works better than `// blah blah` for intellisense, I think.